### PR TITLE
fix(run-work-router): distinguish gh probe failure from not-found

### PR DESCRIPTION
**Pull request by lhpaul**
## Summary

`resolve_token()` in `scripts/development-workflow/run-work-router.sh` probed `gh pr view` and `gh issue view` with `2>/dev/null || true` and treated any empty result as "not found". Any `gh` failure — rate limit, auth expiry, network, GitHub outage — was therefore indistinguishable from a genuinely missing target, and both produced `MODE=ambiguous`, which every bounded command (`/run-item`, `/run-items`, `/run-epic`) treats as a hard stop.

Live reproduction from the issue: `/run-items 1502 1501` stopped reporting both issues unresolvable immediately after a `/run-work` scan exhausted the hourly GraphQL quota, even though both issues were open and had resolved successfully minutes earlier.

## Changes

- `gh_probe()`: captures a gh probe's stdout, stderr, and exit code separately instead of discarding stderr and swallowing the exit code. Restores the caller's original errexit state on return (rather than unconditionally re-enabling `set -e`) so `resolve_token()`'s callers, which wrap the call in `set +e` / `set -e` to inspect `$?`, are not broken by an internal helper silently flipping shell options back on.
- `classify_gh_probe_error()`: classifies a failed probe's stderr into `not_found | rate_limited | auth_failed | network_error | github_unavailable`. Falls back to `not_found` only for gh's own explicit "could not resolve to a PullRequest/Issue" message or an empty stderr (gh's historical behavior for these probes), preserving current behavior for a genuine not-found target.
- `build_probe_fail_reason()` / `gh_rate_limit_reset_info()`: build an actionable `RESOLVE_FAIL_REASON` per failure cause; a rate-limited failure includes the GraphQL quota reset time from `gh api rate_limit` when available.
- New routing mode `MODE_TRACKER_UNAVAILABLE` (`tracker_unavailable`), distinct from `ambiguous`, applied via a new `RESOLVE_FAIL_KIND` at both the single-token and multi-token resolution call sites.

## Testing

New regression tests in `scripts/development-workflow/tests/test-run-work-router.sh` cover:
- Rate-limited probe → `tracker_unavailable`, reason names the cause and includes a reset-time hint.
- Auth-failed probe → `tracker_unavailable`, reason points at `gh auth login`.
- Network-error probe → `tracker_unavailable`.
- Opaque/unrecognized-error (outage-style) probe → `tracker_unavailable` (`github_unavailable` fallback).
- Explicit gh "could not resolve" not-found message on both probes → still `ambiguous` (unchanged).
- Bare non-zero exit with no stderr (pre-existing `999999` fixture) → still `ambiguous` (unchanged).
- A probe failure partway through a multi-token list → `tracker_unavailable`, not masked by an earlier successful resolution.
- Happy path (no probe failure) → unaffected.

All 11 new assertions were confirmed to fail against the pre-fix script (verified via `git stash` of the router change only, keeping the new tests) before the fix was applied. Full local suite: **73/73 passing** (was 58/58 pre-change).

Note: `test-run-work-router.sh` is one of the suites CI does not currently execute (tracked separately in #1537), so this local run is the primary verification.

Fixes #1503

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added a “Tracker unavailable” routing status for rate limits, authentication failures, network issues, outages, and configuration problems.
  - Added actionable stop reasons, including rate-limit reset timing when available.
  - Distinguished genuine missing targets from tracker failures.

- **Bug Fixes**
  - Prevented probe failures from being reported as ambiguous targets.
  - Preserved successful routing for single and multiple targets.

- **Documentation**
  - Updated routing guidance and changelog with the new status, diagnostics, and retry recommendations.

- **Tests**
  - Added coverage for tracker failures, missing targets, multi-target resolution, and successful routing.

### CODE REVIEW

**Comment by lhpaul** [comment]
### Automated Reviewer Loop Summary

**Result:** clean — no blocking findings
**Platforms:** pr-agent (clean), coderabbit (clean)
**Findings:** 0 blocking, 0 suggestions

**Ready reviewer phase:** No net-new blocker was found after the draft GitHub gate.
**Ready-phase platforms:** coderabbit

**Advisory findings (non-blocking):**
- Possible Issue ([view comment](https://github.com/lhpaul/ai-dev-framework-template/pull/1541#issuecomment-5364749454))
- Possible Issue ([view comment](https://github.com/lhpaul/ai-dev-framework-template/pull/1541#issuecomment-5364749454))

*Posted automatically by `pr-review-loop.sh`.*

<details>
<summary>Reviewer-loop history (6 iterations)</summary>


```json
{
  "schema": "reviewer_loop_history.v1",
  "pr_number": 1541,
  "updated_at": "2026-08-21T04:52:01Z",
  "history_status": "available",
  "history_unavailable_reason": "",
  "entries": [
    {
      "iteration": 1,
      "recorded_at": "2026-08-21T03:14:12Z",
      "head_sha": "da3fb6daa82fff7c323b735d47670ae271a76efb",
      "run_id": "auto-1787282039-56578-32758",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": false,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    },
    {
      "iteration": 2,
      "recorded_at": "2026-08-21T03:18:18Z",
      "head_sha": "311d0eb4985b6eee6c42138b35433909024276c7",
      "run_id": "auto-1787282280-83407-20135",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": false,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    },
    {
      "iteration": 3,
      "recorded_at": "2026-08-21T03:23:24Z",
      "head_sha": "311d0eb4985b6eee6c42138b35433909024276c7",
      "run_id": "auto-1787282430-85578-19578",
      "result": "needs_fixes",
      "reason": "",
      "platforms": [
        "pr-agent (clean)",
        "coderabbit (needs_fixes)"
      ],
      "blocking_count": 3,
      "suggestion_count": 1,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": true,
        "net_new_blocker": true,
        "blocking_platform": "coderabbit"
      }
    },
    {
      "iteration": 4,
      "recorded_at": "2026-08-21T03:31:25Z",
      "head_sha": "97384ef7d1f8e0111a4f51a9ed4ee81e85bc712d",
      "run_id": "auto-1787283055-35629-25246",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)",
        "coderabbit (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": true,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    },
    {
      "iteration": 5,
      "recorded_at": "2026-08-21T03:40:50Z",
      "head_sha": "48c7d6bf25cb33b11f86f7b83d9f144c7934c26a",
      "run_id": "auto-1787283509-82336-32624",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)",
        "coderabbit (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": true,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    },
    {
      "iteration": 6,
      "recorded_at": "2026-08-21T04:52:01Z",
      "head_sha": "d33f08196e1822cbc79c601eee1cca8e30bdc6e4",
      "run_id": "auto-1787287904-12401-4766",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)",
        "coderabbit (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": true,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    }
  ]
}
```
</details>

**Comment by lhpaul** [comment]
Addressed both PR-Agent "Possible Issue" findings from the draft review (commit 311d0eb4):

1. **`gh_probe()` errexit ordering**: agreed — moved `PROBE_ERR="$(cat "$err_file")"` and `rm -f "$err_file"` to run before `set -e` is restored, so a failing `cat` can no longer bypass the caller's `set +e` protection.
2. **Empty-stderr fallback masking non-not-found failures**: agreed — `classify_gh_probe_error()` now also takes the probe's exit code and only classifies an empty stderr as `not_found` when the exit code is gh's normal API-error code (`1`). Any other exit code with empty stderr (signal death, OOM kill, launch failure) now falls back to `github_unavailable`.

New regression test (`crash_empty_stderr_nonone_exit_*`, token `888886`) mocks both the PR and issue probes with empty stderr + exit 137, confirmed to fail against the prior commit before this fix (both probes returned `not_found`, resolving to `MODE=ambiguous` instead of `tracker_unavailable`).

Full suite: 75/75 passing.

File: scripts/development-workflow/run-work-router.sh [hunk 1]
```diff
@@ -34,12 +35,14 @@ MODE_REDIRECT_ITEM="redirect_item"
 MODE_REDIRECT_ITEMS="redirect_items"
 MODE_REDIRECT_EPIC="redirect_epic"
 MODE_AMBIGUOUS="ambiguous"
+MODE_TRACKER_UNAVAILABLE="tracker_unavailable"
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply scripts/development-workflow/run-work-router.sh:38]: Fixed: Protocol 96 (canonical routing specification) is updated with the new `tracker_unavailable` mode's routing-modes row, decision-table row, edge cases, and handoff mapping (routing layer version 1.0 -> 1.1), instead of documenting the mode only in the router's own header comment. See the latest commit on this PR.

File: scripts/development-workflow/run-work-router.sh [hunk 2]
```diff
@@ -266,13 +273,180 @@ RESOLVED_KIND=""
...
+  err_file="$(mktemp)"
+  set +e
+  PROBE_OUT="$(gh "$@" 2>"$err_file")"
+  PROBE_EXIT=$?
+  # Capture stderr and clean up the temp file while errexit is still
+  # disabled — a failing `cat` (e.g. the temp file became unreadable) must
+  # not itself trigger `set -e` and terminate the script ahead of the
+  # caller's own `$?` check, once errexit is restored below.
+  PROBE_ERR="$(cat "$err_file" 2>/dev/null)"
+  rm -f "$err_file" 2>/dev/null
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply scripts/development-workflow/run-work-router.sh]: Fixed: gh_probe() now checks mktemp and cat for failure explicitly. If either fails, it sets PROBE_EXIT=125 (a value gh never returns) and a diagnostic PROBE_ERR, instead of letting a resulting empty PROBE_ERR be misread by classify_gh_probe_error() as gh's own empty-stderr not-found case. This mirrors your proposed handling closely (kept the temp-file cleanup after the cat check, matching this function's existing errexit-restore-last ordering).

File: scripts/development-workflow/run-work-router.sh [hunk 3]
```diff
@@ -266,13 +273,180 @@ RESOLVED_KIND=""
...
+    fi
+    return 0
+  fi
+
+  case "$err_lc" in
+    *"could not resolve to a pullrequest"*|*"could not resolve to an issue"*|*"no pull requests found"*|*"no default remote repository has been set"*)
+      echo "not_found"
+      return 0
+      ;;
+  esac
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply scripts/development-workflow/run-work-router.sh:394]: Fixed: removed "no default remote repository has been set" from the not_found pattern list, exactly as suggested — it now falls through to github_unavailable. Added a regression fixture (token 888887, mocking both the PR and issue probes with this message) confirmed to fail against the prior commit before this fix (both probes classified not_found, resolving to MODE=ambiguous instead of tracker_unavailable).

File: CHANGELOG.md [hunk 4]
```diff
@@ -940,6 +940,25 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  as a hard stop with a misdirecting reason. Live reproduction: `/run-items
+  1502 1501` stopped reporting both issues unresolvable immediately after a
+  `/run-work` scan exhausted the hourly GraphQL quota, even though both were
+  open and had resolved successfully minutes earlier. `resolve_token()` now
+  captures each probe's stdout, stderr, and exit code separately (`gh_probe`)
+  and classifies a non-zero exit's stderr (`classify_gh_probe_error`) into
+  `rate_limited`, `auth_failed`, `network_error`, or `github_unavailable`
+  before falling back to `not_found` for gh's own "could not resolve to a
+  PullRequest/Issue" message or an empty stderr — preserving current behavior
+  for a genuine not-found. A probe failure now stops with the distinct
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply CHANGELOG.md]: Fixed: the CHANGELOG entry now states the exit-code condition explicitly — empty stderr only falls back to not_found when the exit code is gh's normal API-error code (1); any other exit code with empty stderr falls back to github_unavailable, matching the implementation.

**Comment by lhpaul** [comment]
Addressed the late-arriving outside-diff finding (docs/workflow/development-workflow/protocols/96-run-work-routing-protocol.md:47-60) — thanks for catching it after the prior fix round.

Went a step further than a docs-only fix: the underlying `github_unavailable` fallback's STOP_REASON text says "retry shortly", which is actively misleading for the "no default remote repository has been set" case — retrying does nothing until `gh repo set-default` is run. So this is now a distinct `local_config_error` classification in the code (not just documented as a github_unavailable sub-case), with its own actionable STOP_REASON. Protocol 96 documents it across the mode description, decision table, edge cases, STOP_REASON field doc, and handoff mapping.

Regression test updated to assert the `gh repo set-default` guidance specifically, confirmed to fail against the prior commit (which produced the generic "retry shortly" wording for this case).

Full suite: 77/77 passing.

**Comment by lhpaul** [comment]
Re: the temp-file rm cleanup finding (run-work-router.sh:331) — partially applied, partially declined with evidence.

**Applied**: added `--` before `"$err_file"` in the `rm -f` call — cheap, harmless defense-in-depth.

**Declined**: overwriting `PROBE_OUT`/`PROBE_ERR`/`PROBE_EXIT` when `rm` fails. Verified against the current code: at that point in `gh_probe()`, those three variables already hold gh's own actual result — including a genuine **successful** resolution (e.g. `PROBE_OUT="OPEN"` for a PR that really exists). The `rm` cleanup runs strictly *after* that capture. If I applied the suggested diff verbatim, a benign filesystem cleanup failure (extremely rare for a private mktemp file the same process just created and read) would silently discard an already-valid successful gh result and misreport it as `tracker_unavailable` — a real regression for `/run-item`/`/run-items`/`/run-epic`, not a fix. The mktemp-failure and cat-failure guards added earlier in this function are correctly placed because those failures happen *before* gh's result is known; this one is different because it happens *after*.

Full suite: 77/77 passing (no behavior change from the `--` addition, so no new assertions).

**Comment by lhpaul** [comment]
@coderabbitai review

**Comment by lhpaul** [comment]
@coderabbitai resume

**Comment by lhpaul** [comment]
@coderabbitai resume

**Comment by lhpaul** [comment]
@coderabbitai resume

File: scripts/development-workflow/run-work-router.sh [hunk 5]
```diff
@@ -266,13 +273,221 @@ RESOLVED_KIND=""
...
+  fi
+  # Best-effort cleanup only. PROBE_OUT/PROBE_ERR/PROBE_EXIT above already
+  # correctly reflect gh's own result (including a genuine successful
+  # resolution) — a failure to delete the now-fully-read temp file must NOT
+  # overwrite them: doing so would discard an already-valid successful probe
+  # result (e.g. a PR that was actually found) and misreport it as a probe
+  # failure over a benign filesystem cleanup blip, unrelated to whether gh
+  # itself succeeded. `--` guards against a pathological temp path starting
+  # with `-`, though mktemp never generates one.
+  rm -f -- "$err_file" 2>/dev/null
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply scripts/development-workflow/run-work-router.sh]: Fixed: a failing rm -f -- "$err_file" now emits a structured 'run-work-router: warning: ...' message to stderr (matching this script's existing warning conventions) instead of being silently swallowed by 2>/dev/null. PROBE_OUT/PROBE_ERR/PROBE_EXIT remain untouched by cleanup outcome, as established in the prior round -- this only adds visibility for diagnosis, no behavior change to the routing decision. 77/77 tests still pass.
